src/core/include/units/bits/prime.h

```diff
@@ -0,0 +1,169 @@
+{
+  std::array<std::size_t, N> primes;
+  primes[0] = 2;
+  for (std::size_t i = 1; i < N; ++i) {
+    primes[i] = primes[i - 1] + 1;
+    while (!is_prime_by_trial_division(primes[i])) {
+      ++primes[i];
+    }
+  }
+  return primes;
+}
+
+template<std::size_t N, typename Callable>
+constexpr void call_for_coprimes_up_to(std::size_t n, const std::array<std::size_t, N>& basis, Callable&& call)
+{
+  for (std::size_t i = 0u; i < n; ++i) {
+    if (std::apply([&i](auto... primes) { return ((i % primes != 0) && ...); }, basis)) {
+      call(i);
+    }
+  }
+}
+
+template<std::size_t N>
+constexpr std::size_t num_coprimes_up_to(std::size_t n, const std::array<std::size_t, N>& basis)
+{
+  std::size_t count = 0u;
+  call_for_coprimes_up_to(n, basis, [&count](auto) { ++count; });
+  return count;
+}
+
+template<std::size_t ResultSize, std::size_t N>
+constexpr auto coprimes_up_to(std::size_t n, const std::array<std::size_t, N>& basis)
+{
+  std::array<std::size_t, ResultSize> coprimes;
+  std::size_t i = 0u;
+
+  call_for_coprimes_up_to(n, basis, [&coprimes, &i](std::size_t cp) { coprimes[i++] = cp; });
+
+  return coprimes;
+}
+
+template<std::size_t N>
+constexpr std::size_t product(const std::array<std::size_t, N>& values)
+{
+  std::size_t product = 1;
+  for (const auto& v : values) {
+    product *= v;
+  }
+  return product;
+}
+
+// A configurable instantiation of the "wheel factorization" algorithm [1] for prime numbers.
+//
+// Instantiate with N to use a "basis" of the first N prime numbers.  Higher values of N use fewer trial divisions, at
+// the cost of additional space.  The amount of space consumed is roughly the total number of numbers that are a) less
+// than the _product_ of the basis elements (first N primes), and b) coprime with every element of the basis.  This
+// means it grows rapidly with N.  Consider this approximate chart:
+//
+//   N | Num coprimes | Trial divisions needed
+//   --+--------------+-----------------------
+//   1 |            1 |                 50.0 %
+//   2 |            2 |                 33.3 %
+//   3 |            8 |                 26.7 %
+//   4 |           48 |                 22.9 %
+//   5 |          480 |                 20.8 %
+//
+// Note the diminishing returns, and the rapidly escalating costs.  Consider this behaviour when choosing the value of N
+// most appropriate for your needs.
+//
+// [1] https://en.wikipedia.org/wiki/Wheel_factorization
+template<std::size_t BasisSize>
+struct wheel_factorizer {
+  static constexpr auto basis = first_n_primes<BasisSize>();
+  static constexpr std::size_t wheel_size = product(basis);
+  static constexpr auto coprimes_in_first_wheel =
+    coprimes_up_to<num_coprimes_up_to(wheel_size, basis)>(wheel_size, basis);
+
+  static constexpr std::size_t find_first_factor(std::size_t n)
+  {
+    for (const auto& p : basis) {
+      if (const auto k = first_factor_maybe(n, p)) {
+        return *k;
+      }
+    }
```

Comment:
Please use `detail::find_if()` and provide a TODO comment. See more in: https://github.com/mpusz/units/commit/e1f7266b517c04644b666af6aef3e65216b28c56.

Reply:
If I do that, how will I know which value to return without calling `first_factor_maybe()` again?  Granted, that's not very expensive, but it seems a little inelegant.

In more detail: I guess this would look like:

```cpp
const auto basis_iter = find_if(basis.begin(), basis.end(), [n](auto iter) { return first_factor_maybe(n, *iter); });
if (basis_iter) {
  return *first_factor_maybe(n, *basis_iter);
}
```

Are we sure that's better?

In the existing code, I can see the nuisance of repeating the `if` block four times.  The best solution that occurs to me would be to separate the value generation from the evaluation---maybe have some kind of iterator interface which returns the basis primes in order, then first-wheel coprimes (except 1), then coprimes in order for each wheel indefinitely.  I assume that's doable with something like ranges or coroutines, although I'm not very well versed in either.  I think it would be fine to leave as a future refactoring.

That said---I did realize in grappling with this comment that one of the four repetitions was redundant!  It stems from my earlier mistake of using primes; 1 is not a prime, but it _is_ coprime, so we were checking it twice per wheel.  That's fixed now.

Let me know your thoughts on how to proceed here: status quo, or a `find_if` solution.

Reply:
Also, to be clear about why "which value to return" is even an issue: `first_factor_maybe` can return either `k` itself, or `n` (if `k * k` exceeds `n`).

Reply:
Right, initially I thought it will be easier to simplify... Sorry, for causing confusion.

Reply:
However, it looks like the same pattern in all cases so maybe a custom algorithm can be provided to cover those?

Reply:
I want that too, but I think `first_factor_maybe()` _is_ that algorithm---at least, as much of it as I was able to pull out. I think the next level of refactoring would be to separate out the trial number generation, probably placing it behind some kind of iterator interface. Then we could just keep taking values until we find one that works, and we'd have only a single call to `first_factor_maybe()` (or more likely, we'd just get rid of that function at that point).

If you have a ready made solution for this, I'd love to add it! Otherwise, I think the code is reasonably clean and efficient for a first pass.